drivers: adc: nrfx_saadc: Add support for SAADC internal sampling timer

jaz1-nordic opened the pull request:
The SAMPLERATE register can be used as a local timer instead of triggering individual SAMPLE tasks. When SAMPLERATE.MODE is set to Timers, it is sufficient to trigger SAMPLE task only once in order to start the SAADC and triggering the STOP task will stop sampling.
The SAMPLERATE.CC field controls the sample rate.

The SAMPLERATE timer should not be combined with SCAN mode and only one channel should be enabled when using the internal timer.

bjarki-andreasen commented on drivers/adc/adc_context.h:
```diff
@@ -115,6 +115,7 @@ static inline void adc_context_request_next_sampling(struct adc_context *ctx)
 }
 
 #ifdef ADC_CONTEXT_USES_KERNEL_TIMER
+#ifndef ADC_DRIVER_USES_INTERNAL_TIMER
```
~This config is not defined~, and is ~also~ not useful from what I see, the APIs here are static inline, they don't take up resources if not used.

bjarki-andreasen commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -5,6 +5,8 @@
  */
 
 #define ADC_CONTEXT_USES_KERNEL_TIMER
+#define ADC_DRIVER_USES_INTERNAL_TIMER
```
Given this is always defined, why define it? just remove this definition and all the ifdefs accompanying it :)

bjarki-andreasen commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -104,15 +106,26 @@ struct driver_data {
 	ADC_CONTEXT_INIT_SYNC(m_data, ctx),
 	.mem_reg = DMM_DEV_TO_REG(DT_NODELABEL(adc)),
+#ifdef ADC_DRIVER_USES_INTERNAL_TIMER
+	.enable_internal_timer = DT_INST_PROP(0, enable_internal_timer),
```
This does not seem like an option we should present to the user, and definately not in the devicetree. If we can use the internal timer, just use it :) Is there some reason not to use it?

bjarki-andreasen commented on dts/bindings/adc/nordic,nrf-saadc.yaml:
```diff
@@ -14,6 +14,16 @@ properties:
   interrupts:
     required: true
 
+  enable-internal-timer:
+    type: boolean
+    description: |
+      Enable the internal timer for the SAADC to improve interval accuracy.
+      When enabled, the SAADC will utilize its built-in local timer for sampling.
+      If disabled, the SAADC will rely on the kernel timer.
+      WARNING: The internal timer is only applicable for sequences with a single channel.
+               It can be used only for intervals between consecutive samplings ranging
+               from 500 ns to 128 us.
+
```
This really should be handled internally in the driver, if a user request a sequence with only 1 channel, and the interval is within what the internal timer supports, use the internal timer. If a user requests a sequence with multiple channels, or unsupported interval, use kernel timer. A LOG_INF() can be added to inform the user which timer was used to let users know why the intervals may be less accurate than expected.

bjarki-andreasen reviewed (changes requested):
Adding support for the internal timer is great, but I believe it can be simplified a bit, especially for the user's perspective

bjarki-andreasen reviewed (dismissed):
Looks good!

jaz1-nordic commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -392,24 +385,58 @@ static void adc_context_start_sampling(struct adc_context *ctx)
 {
-	if (!repeat) {
-#if defined(ADC_BUFFER_IN_RAM)
+	if (!repeat && !m_data.internal_timer_enabled) {
```
Done

jaz1-nordic commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -97,36 +100,26 @@ BUILD_ASSERT((NRF_SAADC_AIN0 == NRF_SAADC_INPUT_AIN0) &&
-
-#if defined(ADC_BUFFER_IN_RAM)
-	void *samples_buffer;
+	void *mem_reg;
```
#90751 is already approved. Waiting for merge.

bjarki-andreasen commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -5,6 +5,8 @@
  */
 
 #define ADC_CONTEXT_USES_KERNEL_TIMER
+#define ADC_DRIVER_USES_INTERNAL_TIMER
```
From what I can see, this option makes the ADC context not control the timer, only define and init it, for the driver to then manage manually, could you add a description of the purpose of this define in adc_context.h ? also, maybe rename it `ADC_DRIVER_MANAGES_KERNEL_TIMER` or something if my understanding is correct, otherwise its not clear that this relates to same timer as `ADC_CONTEXT_USES_KERNEL_TIMER`

jaz1-nordic replied:
@bjarki-andreasen please review. I removed the defines and moved the timer support directly to the driver according to the instructions from `adc_context.h`:
See line 33:
```
 * If a given driver uses some dedicated hardware timer to trigger consecutive
 * samplings, it should implement also the following two functions. Otherwise,
 * it should define the ADC_CONTEXT_USES_KERNEL_TIMER macro to enable parts of
 * this module that utilize a standard kernel timer.
```

magp-nordic reviewed (dismissed):
LGTM, just left some notes with possible improvements for future.

bjarki-andreasen commented:
ping @anangl

jaz1-nordic committed:
drivers: adc: nrfx_saadc: Add support for SAADC internal sampling timer

The SAMPLERATE register can be used as a local timer instead
of triggering individual SAMPLE tasks. When SAMPLERATE.MODE is set
to Timers, it is sufficient to trigger SAMPLE task only once in order
to start the SAADC and triggering the STOP task will stop sampling.
The SAMPLERATE.CC field controls the sample rate.

The SAMPLERATE timer should not be combined with SCAN mode and
only one channel should be enabled when using the internal timer.

Signed-off-by: Jakub Zymelka <jakub.zymelka@nordicsemi.no>

jaz1-nordic commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -490,31 +529,68 @@ static int check_buffer_size(const struct adc_sequence *sequence, uint8_t active
-				*sample = 0;
+	for (uint16_t channel_bit = BIT(0); channel_bit <= single_ended_channels;
+	     channel_bit <<= 1) {
+		if (!(channel_bit & selected_channels)) {
```
I simplified this.
